numba/compiler.py

```diff
@@ -680,48 +694,94 @@ def stage_cleanup(self):
+                pm.add_stage(self.stage_preserve_ir,
+                             "preserve IR for fallback")
+            pm.add_stage(self.stage_generic_rewrites, "nopython rewrites")
+        pm.add_stage(self.stage_inline_pass,
+                     "inline calls to locally defined closures")
+
+    def add_typing_stage(self, pm):
+        """Add the type-inference stage necessary for nopython mode.
+        """
+        pm.add_stage(self.stage_nopython_frontend, "nopython frontend")
+        pm.add_stage(self.stage_annotate_type, "annotate type")
 
-        if self.status.can_fallback or self.flags.force_pyobject:
-            pm.create_pipeline("object")
-            if self.func_ir is None:
-                pm.add_stage(self.stage_analyze_bytecode, "analyzing bytecode")
-            pm.add_stage(self.stage_process_ir, "processing IR")
-            pm.add_stage(self.stage_objectmode_frontend, "object mode frontend")
-            pm.add_stage(self.stage_annotate_type, "annotate type")
-            pm.add_stage(self.stage_objectmode_backend, "object mode backend")
-            pm.add_stage(self.stage_cleanup, "cleanup intermediate results")
+    def add_optimization_stage(self, pm):
+        """Add optimization stages.
+        """
+        if self.flags.auto_parallel.enabled:
+            pm.add_stage(self.stage_pre_parfor_pass,
+                         "Preprocessing for parfors")
+        if not self.flags.no_rewrites:
+            pm.add_stage(self.stage_nopython_rewrites, "nopython rewrites")
+        if self.flags.auto_parallel.enabled:
+            pm.add_stage(self.stage_parfor_pass, "convert to parfors")
+
+    def add_lowering_stage(self, pm):
+        """Add the lowering (code-generation) stage for nopython-mode
+        """
+        pm.add_stage(self.stage_nopython_backend, "nopython mode backend")
 
-        if self.status.can_giveup:
-            pm.create_pipeline("interp")
-            pm.add_stage(self.stage_compile_interp_mode, "compiling with interpreter mode")
-            pm.add_stage(self.stage_cleanup, "cleanup intermediate results")
+    def add_cleanup_stage(self, pm):
+        """Add the clean-up stage to remove intermediate results.
+        """
+        pm.add_stage(self.stage_cleanup, "cleanup intermediate results")
 
+    def define_nopython_pipeline(self, pm, name='nopython'):
+        """Add the nopython-mode pipeline to the pipeline manager
+        """
+        pm.create_pipeline(name)
+        self.add_preprocessing_stage(pm)
+        self.add_pre_typing_stage(pm)
+        self.add_typing_stage(pm)
+        self.add_optimization_stage(pm)
+        self.add_lowering_stage(pm)
+        self.add_cleanup_stage(pm)
+
+    def define_objectmode_pipeline(self, pm, name='object'):
+        """Add the object-mode pipeline to the pipeline manager
+        """
+        pm.create_pipeline(name)
+        self.add_preprocessing_stage(pm)
+        pm.add_stage(self.stage_objectmode_frontend,
```

Comment:
The inversion of concern here is a bit weird, `self.add_stage_foo(pm)` vs. `pm.add_stage(foo)`. Perhaps another refactor at a later date?

Reply:
I agree.  That are still a lot of things that need cleaning up.